File: chapter_0/binary_search/binary_search/binary_search.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-27 16:40:17
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-01-27 17:43:00
 */

#include <iostream>
#include <vector>

using namespace std;

int binary_search(const vector<int>& nums, const int& target) {
  int left = 0;
  int right = nums.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] == target)
      return mid;
    else if (nums[mid] < target)
      left = mid + 1;
    else if (nums[mid] > target)
      right = mid - 1;
  }
  return -1;
}

int left_bound(const vector<int>& nums, const int& target) {
  int left = 0;
  int right = nums.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] == target)
      right = mid - 1;
    else if (nums[mid] < target)
      left = mid + 1;
    else if (nums[mid] > target)
      right = mid - 1;
  }
  if (left >= nums.size() || nums[left] != target) return -1;
  return left;
}

int right_bound(const vector<int>& nums, const int& target) {
  int left = 0;
  int right = nums.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] == target)
      left = mid + 1;
    else if (nums[mid] < target)
      left = mid + 1;
    else if (nums[mid] > target)
      right = mid - 1;
  }
  if (right < 0 || nums[right] != target) return -1;
  return right;
}

int main() {
  vector<int> nums = {1, 2, 3, 4, 5};
  int target = 3;
  std::cout << binary_search(nums, target) << std::endl;

  nums = {2, 2, 2, 2, 2};
  target = 2;
  std::cout << left_bound(nums, target) << std::endl;
  std::cout << right_bound(nums, target) << std::endl;

  nums = {1, 2, 2, 2, 3};
  std::cout << left_bound(nums, target) << std::endl;
  std::cout << right_bound(nums, target) << std::endl;

  target = 4;
  std::cout << left_bound(nums, target) << std::endl;
  std::cout << right_bound(nums, target) << std::endl;

  return 0;
}

File: chapter_0/binary_search/binary_search/binary_search.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-27 16:40:23
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-01-27 17:31:42
 */

fn binary_search(nums: &Vec<i32>, target: &i32) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = nums.len() as i32 - 1;
    let mut mid: i32;

    while left <= right {
        mid = left + (right - left) / 2;
        if nums[mid as usize] == *target {
            return mid;
        } else if nums[mid as usize] < *target {
            left = mid + 1;
        } else if nums[mid as usize] > *target {
            right = mid - 1;
        }
    }
    -1
}

fn left_bound(nums: &Vec<i32>, target: &i32) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = nums.len() as i32 - 1;
    let mut mid: i32;

    while left <= right {
        mid = left + (right - left) / 2;
        if nums[mid as usize] == *target {
            right = mid - 1;
        } else if nums[mid as usize] < *target {
            left = mid + 1;
        } else if nums[mid as usize] > *target {
            right = mid - 1;
        }
    }
    if left >= nums.len() as i32 || nums[left as usize] != *target {
        return -1;
    }
    left
}

fn right_bound(nums: &Vec<i32>, target: &i32) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = nums.len() as i32 - 1;
    let mut mid: i32;

    while left <= right {
        mid = left + (right - left) / 2;
        if nums[mid as usize] == *target {
            left = mid + 1;
        } else if nums[mid as usize] < *target {
            left = mid + 1;
        } else if nums[mid as usize] > *target {
            right = mid - 1;
        }
    }
    if right < 0 || nums[right as usize] != *target {
        return -1;
    }

    right
}

fn main() {
    let nums: Vec<i32> = vec![1, 2, 3, 4, 5];
    let target: i32 = 3;
    println!("{}", binary_search(&nums, &target));

    let nums: Vec<i32> = vec![2, 2, 2, 2, 2];
    let target: i32 = 2;
    println!("{}", left_bound(&nums, &target));
    println!("{}", right_bound(&nums, &target));

    let nums: Vec<i32> = vec![1, 2, 2, 2, 3];
    println!("{}", left_bound(&nums, &target));
    println!("{}", right_bound(&nums, &target));

    let target: i32 = 4;
    println!("{}", left_bound(&nums, &target));
    println!("{}", right_bound(&nums, &target));
}
